sb_vision/native/apriltag/__init__.py

```diff
@@ -1 +1,98 @@
 """AprilTag native utilities."""
+
+from PIL import Image
+
+from ._apriltag import ffi, lib
+
+
+__all__ = ('AprilTagLibrary',)
+
+
+class AprilTagLibrary:
+    def __init__(self, image_size):
+        """
+        Initialise the AprilTag library.
+
+        This means creating and configuring the detector, which populates a
+        number of tables in memory.
+        """
+        self._image_size = image_size
+
+        self._detector = lib.apriltag_detector_create()
+        """
+        apriltag_detector_t* td,
+        float decimate,
+          default: 1.0, "Decimate input image by this factor"
+        float sigma,
+          default: 0.0, "Apply low-pass blur to input; negative sharpens"
+        int refine_edges,
+          default: 1, "Spend more time trying to align edges of tags"
+        int refine_decode,
+          default: 0, "Spend more time trying to decode tags"
+        int refine_pose
+          default: 0, "Spend more time trying to find the position of the tag"
+        """
+
+        lib.apriltag_init(self._detector, 1.0, 0.0, 1, 0, 0)
+
+        # TODO: keeping this around for the lifetime of the library feels like
+        # it's an optiisation rather than actually required. Given that it makes
+        # the instance non-thread-safe, work out whether it's desirable and/or
+        # there are other reasons the instance isn't thread-safe anyway.
+        self._working_image = lib.image_u8_create_stride(
+            image_size[0],
+            image_size[1],
+            image_size[0],
+        )
+
+    def __del__(self):
+        """Deinitialise the library."""
+
+        if self._detector:
+            lib.apriltag_detector_destroy(self._detector)
+
+        if self._working_image:
+            lib.image_u8_destroy(self._working_image)
```

Comment:
I'm not wild about `__del__` for resource management, because Python makes no guarantees about when (or even whether) `__del__` is called.

Would be up for a context manager/explicit `close` to do this.

Reply:
c97a296